Vostok.Hosting/Components/Log/EventLevelCounterFactory.cs

```diff
@@ -0,0 +1,22 @@
+﻿using JetBrains.Annotations;
+
+namespace Vostok.Hosting.Components.Log
+{
+    [PublicAPI]
+    public class EventLevelCounterFactory
+    {
+        private readonly LevelCountingLog baseLog;
```

Comment:
It feels like tight coupling of `EventLevelCounterFactory` and `LevelCountingLog` made the code tangled and rather complex (especially in `Logs` class).

It could be simpler if `EventLevelCounterFactory` knew nothing about log decorators and just kept an immutable list of currently registered counters. `LevelCountingLog` would then just accept a `Func<IEnumerable<EventLevelCounter>>`.

Pros of this approach:
- No chance of creating potentially long nested chains of decorators (this may harm performance);
- Easy to enable removal of no longer needed counters since `EventLevelCounterFactory` manages them exclusively;
- There would be no need to create the factory during log customization in `Logs`;

Also, there is a race condition here: two concurrent CreateCounter calls may lead to loss of one decorator.